script/copy_deduplicate

```diff
@@ -19,44 +19,42 @@ from uuid import uuid4
 
 from google.cloud import bigquery
 
-QUERY_DAY_BASE = """
-  -- A single day of a live ping table.
-  base AS (
-  SELECT
-    *
-  FROM
-    `{source_table_spec}`
-  WHERE
-    DATE(submission_timestamp) = @submission_date ),
-"""
-
-QUERY_HOUR_BASE = """
-  -- Document_ids already seen in previous hours on this day
-  previous_hours_document_ids AS (
+QUERY_BASE = """
+  -- Distinct document_ids and their minimum submission_timestamp today
+  -- not including document_ids that only occur on or after @end_time
+  distinct_document_ids AS (
   SELECT
-    DISTINCT document_id
+    document_id,
+    MIN(submission_timestamp) AS submission_timestamp
   FROM
-    `{source_table_spec}`
+    `{live_table}`
   WHERE
-    DATE(submission_timestamp) = DATE(@submission_hour)
-    AND submission_timestamp < @submission_hour),
-  -- A single hour of a live ping table.
+    DATE(submission_timestamp) = DATE(@start_time)
+    AND submission_timestamp < @end_time
+  GROUP BY
+    document_id),
+  -- A single slice of a live ping table.
   base AS (
   SELECT
     *
   FROM
-    `{source_table_spec}`
+    `{live_table}`
+  JOIN
+    distinct_document_ids
+  USING
+    -- Retain only the first seen documents for each ID, according to timestamp.
+    (document_id, submission_timestamp)
```

Comment:
Most duplicates come from clients sending a message multiple times, meaning they get distinct timestamps. But we also have a smaller class of duplicates where timestamps are equal, and it looks like this query is vulnerable to allowing multiple rows with the same document_id for this class of duplicates.

Reply:
I was experimenting with this yesterday and came up with the following to handle counting the number of occurrences of each timestamp and to do a final window function over just that smaller class of duplicates:

```
WITH
  base AS (
  SELECT
    *
  FROM
    `moz-fx-data-shared-prod.telemetry_stable.main_v4`
  WHERE
    DATE(submission_timestamp) = '2019-09-23' ),
  id_time_dupes AS (
  SELECT
    document_id,
    submission_timestamp,
    COUNT(*) AS occurrences
  FROM
    base
  GROUP BY
    document_id,
    submission_timestamp ),
  id_dupes AS (
  SELECT
    document_id,
    ARRAY_AGG(submission_timestamp
    ORDER BY
      submission_timestamp)[
  OFFSET
    (0)] AS submission_timestamp,
    ARRAY_AGG(occurrences
    ORDER BY
      submission_timestamp)[
  OFFSET
    (0)] AS occurrences
  FROM
    id_time_dupes
  GROUP BY
    document_id
  HAVING
    COUNT(*) > 1 ),
  --
  -- We can directly include all singly-occurring documents in the final
  -- result set without incurring the expense of sorting in a window function.
  singly_occurring_documents AS (
  SELECT
    base.*
  FROM
    base
  LEFT JOIN
    id_dupes
  USING
    (document_id)
  WHERE
    id_dupes.document_id IS NULL),
  earliest_dupes AS (
  SELECT
    base.*
  FROM
    base
  JOIN
    id_dupes
  USING
    (document_id,
      submission_timestamp)
  WHERE
    occurrences = 1 ),
  --
  -- For documents that appear multiple times, we have to order them by
  -- assigning a row number.
  numbered_duplicates AS (
  SELECT
    base.*,
    ROW_NUMBER() OVER (PARTITION BY document_id ORDER BY submission_timestamp) AS _n
  FROM
    base
  JOIN
    id_dupes
  USING
    (document_id,
      submission_timestamp)
  WHERE
    occurrences > 1 ),
  --
  -- We retain only the first seen document for each ID, according to timestamp.
  deduplicated_documents AS (
  SELECT
    * EXCEPT (_n)
  FROM
    numbered_duplicates
  WHERE
    _n = 1 )
  --
SELECT
  *
FROM
  singly_occurring_documents
UNION ALL
SELECT
  *
FROM
  earliest_dupes
UNION ALL
SELECT
  *
FROM
  deduplicated_documents
```

Reply:
Looks like the later queries handle this for us, so I'm satisfied.